[storage] Envelope v2

muhamadazmy opened the pull request:
[storage] Envelope v2

# Summary
- Separate header, keys, and other envelope related
information from the actual payload of the envelope
- Ability to decode envelope header without decoding the full envelope
- Support different encoding for the payload (records) (flexbuffers, or bilrost)
- Delay decoding of payload until needed

# TODO:
- [x] Decode v1 envelope into v2
- [ ] Use Envelope v2 in code
- [ ] Convert from v2 to v1 (to write)

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/3853).
* #4058
* __->__ #3853
* #3852

muhamadazmy commented on crates/types/src/message.rs:
```diff
@@ -29,3 +29,11 @@ pub enum AckKind {
+    pub index: MessageIndex,
+}
+
+bilrost_storage_encode_decode!(MessageIndexRecrod);
```
Agree! As for the partition-id this record message was mapped 1:1 from v1 without much thought about the payload it carries. I think the reason it doesn't have the partition id is that that's part of the envelope Header (destination)

muhamadazmy commented on crates/types/src/storage.rs:
```diff
@@ -187,7 +189,7 @@ macro_rules! flexbuffers_storage_encode_decode {
             {
                 $crate::storage::decode::decode_serde(buf, kind).map_err(|err| {
-                    ::tracing::error!(%err, "{} decode failure (decoding {})", kind, stringify!($name));
+                    $crate::storage::tracing::error!(%err, "{} decode failure (decoding {})", kind, stringify!($name));
```
Using this macro in a crate that doesn't have `tracing` as a dependency will fail. I had this issue when i was trying to use the macro in the `wal-protocol` crate which does not depend on tracing. 

So I had 2 options:
- add `tracing` as a dependency to the crate. Only for the sake of this macro. The dependency will not be "used" in the crate anywhere else.
- use a better macro hygiene and use tracing from the declaring crate. Which what is done here.

This way users of the macros will not need to pull extra dependencies (tracing in this case)

muhamadazmy commented on crates/types/src/storage.rs:
```diff
@@ -196,6 +198,40 @@ macro_rules! flexbuffers_storage_encode_decode {
+                &self,
+                buf: &mut ::bytes::BytesMut,
+            ) -> Result<(), $crate::storage::StorageEncodeError> {
+                bytes::BufMut::put(buf, $crate::storage::encode::encode_bilrost(self));
```
Not all bilrost encode variants (for example encode_contiguous which is used here) can make use of  the `BufMut`. But we can pass it anyway and leave the implementation of the encode_bilrost function to decide what to do.

muhamadazmy commented on crates/wal-protocol/src/v2.rs:
```diff
@@ -0,0 +1,824 @@
+    record! {
+        @name=PurgeJournal,
+        @kind=RecordKind::PurgeJournal,
+        @payload=PurgeInvocationRequest
```
That's what v1 does. Yes.

AhmedSoliman commented on crates/wal-protocol/src/v2.rs:
```diff
@@ -0,0 +1,777 @@
+/// and serialization details required to interpret the payload.
+#[derive(Debug, Clone, bilrost::Message)]
+pub struct Header {
+    #[bilrost(1)]
```
Any particular reason why those tags are not 1,2,3?

muhamadazmy committed:
Move Envelope to v1

muhamadazmy committed:
[storage] Envelope v2

# Summary
- Separate header, keys, and other envelope related
information from the actual payload of the envelope
- Ability to decode envelope header without decoding the full envelope
- Support different encoding for the payload (records) (flexbuffers, or bilrost)
- Delay decoding of payload until needed

# TODO:
- [x] Decode v1 envelope into v2
- [ ] Use Envelope v2 in code
- [ ] Convert from v2 to v1 (to write)
